entrypoint/ocp/builder.go

```diff
@@ -0,0 +1,168 @@
+	}
+
+	// Init the API Client for k8s itself
+	// The API client requires that the pod/buildconfig include a service account.
+	k8sAPIErr := k8sAPIClient()
+	if k8sAPIErr != nil && k8sAPIErr != ErrNotInCluster {
+		log.Errorf("Failed to initalized Kubernetes in cluster API client: %v", k8sAPIErr)
+		return nil, k8sAPIErr
+	}
+
+	// Init the OpenShift Build API Client.
+	buildAPIErr := ocpBuildClient()
+	if buildAPIErr != nil && buildAPIErr != ErrNoOCPBuildSpec {
+		log.Errorf("Failed to initalized the OpenShift Build API Client: %v", buildAPIErr)
+		return nil, buildAPIErr
+	}
+
+	// Builder requires either a Build API Client or Kuberneres Cluster client.
+	if k8sAPIErr != nil && buildAPIErr != nil {
+		return nil, ErrInvalidOCPMode
+	}
+
+	if _, err := os.Stat(cosaSrvDir); os.IsNotExist(err) {
+		return nil, fmt.Errorf("Context dir %q does not exist", cosaSrvDir)
+	}
+	// Finally check the API client
+	return &v, nil
+}
+
+// PrepareEnv setups the COSA build environment.
+func (o *Builder) PrepareEnv() error {
+	if err := os.Chdir(cosaSrvDir); err != nil {
+		return fmt.Errorf("Failed to switch to context dir: %s: %v", cosaSrvDir, err)
+	}
+
+	// Load secrets directly from the Kubernetes API that are
+	// are "well-known" secrets.
+	if err := kubernetesSecretsSetup(cosaSrvDir); err != nil {
+		log.Errorf("Failed to setup Service Account Secrets: %v", err)
+	}
+
+	// Read setup the secrets locally.
+	if err := buildSecretsSetup(cosaSrvDir); err != nil {
+		log.Errorf("Failed to setup OCP Build Secrets: %v", err)
+	}
+
+	// Extract any binary sources first. If a binary payload is delivered,
+	// then blindly execute any script ending in .cosa.sh
+	bin, err := extractInputBinary(cosaSrvDir)
+	if err != nil {
+		return err
+	}
+
+	// Locate the job spec.
+	jsF := filepath.Join(cosaSrvDir, rhjobspec.DefaultJobSpecFile)
+	js, err := rhjobspec.JobSpecFromFile(jsF)
+	if err != nil {
+		o.JobSpec = js
+	}
+
+	// If there is no binary payload, then init COSA
+	// With OCP its either binary _or_ source.
+	if !bin {
+		if err := cosaInit(cosaSrvDir); err != ErrNoSourceInput {
+			return err
+		}
+		log.Info("No source input, relying solely on envVars...this won't end well.")
+	}
+
+	return nil
+}
+
+// Exec executes the command using the closure for the commands
+func (o *Builder) Exec(f func(args []string) error) error {
+	curD, _ := os.Getwd()
+	defer os.Chdir(curD)
+	if err := os.Chdir(cosaSrvDir); err != nil {
+		return err
+	}
+
+	// By default, binary payloads override all envvars directives.
+	// OCP does not allow envVars with binary builds.
+	if apiBuild.Spec.Source.Binary != nil {
+		scripts, err := filepath.Glob("*.cosa.sh")
+		if err != nil {
+			return err
+		}
+		return f(scripts)
+	}
+
+	// If there are no binary payloads, write then envVars to
+	// to a temporary file, and use that for execution.
+	tmpf, err := ioutil.TempFile("", "cosa-")
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tmpf.Name())
+	content := "#!/bin/bash\n%s\n"
```

Comment:
[strict mode](http://redsymbol.net/articles/unofficial-bash-strict-mode/)?

Reply:
Yeah, I'll do that in a follow-up. I thought about it. But TBH, I'm wanting to fix up this interface. Its just a starting point. And I'm not settled that this is the interface I want.

Reply:
Fixed.